Extended Tour Documentation (#614)
open-component-model/ocm: examples/lib/tour/docsrc/03-working-with-credentials/README.md

and credential cosunmers in a technology-agnostic way.

It covers four basic scenarios:
- [`basic`](01-using-credentials.go) Writing to a repository with directly specified credentials.
- [`context`](02-basic-credential-management.go) Using credentials via the credential management to publish a component version.
- [`read`](02-basic-credential-management.go) Read the previously created component version using the credential management.
- [`credrepo`](03-credential-repositories.go) Providing credentials via credential repositories.

## Running the example

You can just call the main program with some config file option (`--config <file>`) and the name of the scenario.
The config file should have content similar to:

```yaml
repository: ghcr.io/mandelsoft/ocm
username:
password:
```

Set your favorite OCI registry and don't forget to add the repository prefix for your OCM repository hosted in this registry.

## Walkthrough

### Writing to a repository with directly specified credentials.

As usual, we start with getting access to an OCM context
object.

```go
{{include}{../../02-composing-a-component-version/01-basic-componentversion-creation.go}{default context}}
```

> So far, we just used memory or filesystem based

Reviewer: ```suggestion
So far, we just used memory or file system based
```

Author: see above

OCM repositories to create component versions.
If we want to store something in a remotely accessible
repository typically some credentials are required
for write access.

The OCM library uses a generic abstraction for credentials.
It is just set of properties. To offer various credential sources

> There is an interface `credentials.Credentials` provided,

Author: done

whose implementations provide access to those properties.
A simple property based implementation is `credentials.DirectCredentials.

The most simple use case is to provide the credentials
directly for the repository access creation.
The example config file provides such credentials
for an OCI registry.

```go
{{include}{../../03-working-with-credentials/01-using-credentials.go}{new credentials}}
```

Now, we can use the OCI repository access creation from the [first tour](../01-getting-started/README.md#walkthrough),
but we pass the credentials as additional parameter.

> To give you the chance to specify your own registry the URL

Reviewer: ```suggestion
To give you the chance to specify your own registry, the URL
```

Author: done

is taken from the config file.

```go
{{include}{../../03-working-with-credentials/01-using-credentials.go}{repository access}}
```

If registry name and credentials are fine, we should be able
now to add a new component version to this repository using the coding
from the previous examples, but now we use a public repository, instead

> of a memory or filesystem based one. This coding is in function `addVersion`

Reviewer: ```suggestion
of a memory or file system based one. This coding is in function `addVersion`
```

Author: done

in `common.go` (It is shared by the other examples, also).

```go
{{include}{../../03-working-with-credentials/common.go}{create version}}
```

In contract to our [first tour](../01-getting-started/README.md#walkthrough)
we cannot list components, here.
OCI registries do not support component listers, therefore we

> just lookup the actually added version to verify the result.

Reviewer: ```suggestion
just look up the actually added version to verify the result.
```

Author: done

```go
{{include}{../../03-working-with-credentials/01-using-credentials.go}{lookup}}
```

The coding for `describeVersion` is similar to the one shown in the [first tour]({{describe-version}}).

{{using-cred-management}}
### Using the Credential Management

Passing credentials directly at the repository
is fine, as long only the component version
will be accessed. But as soon as described
resource content will be read, the required
credentials and credential types are dependent
on the concrete component version, because
it might contain any kind of access method
referring to any kind of resource repository
type.

To solve this problem of passing any set
of credentials the OCM context object is

> used to store credentials. This handled

Reviewer: ```suggestion
used to store credentials. This is handled
```

Author: done

by a sub context, the *Credentials context*.

As usual, we start with the default OCM context.

```go
{{include}{../../03-working-with-credentials/02-basic-credential-management.go}{default context}}
```

It is now used to gain access to the appropriate
credential context.

```go
{{include}{../../03-working-with-credentials/02-basic-credential-management.go}{cred context}}
```

The credentials context brings together
providers of credentials, for example a
vault or a local docker/config.json
and credential consumers like GitHub or
OCI registries.
It must be able to distinguish various kinds
of consumers. This is done by identifying
a dedicated consumer with a set of properties
called `credentials.ConsumerId`. It consists
at least of a consumer type property and a
consumer type specific set of properties
describing the concrete instance of such
a consumer, for example an OCI artifact in
an OCI registry is identified by a host and
a repository path.

A credential provider like a vault just provides

> named credential set and typically does not

Reviewer: ```suggestion
named credential sets and typically does not
```

Author: done

know anything about the use case for these sets.

> The task of the credential context is now to

Reviewer: ```suggestion
The task of the credential context is to
```

Author: done

provide credentials for a dedicated consumer.
Therefore, it maintains a configurable
mapping of credential sources (credentials in
a credential repository) and a dedicated consumer.

> This mapping defines a usecase, also based on

Reviewer: ```suggestion
This mapping defines a use case, also based on
```

Author: done

a property set and dedicated credentials.
If credentials are required for a dedicated
consumer, it matches the defined mappings and
returned the best matching entry.

> Matching? Let's take GitHub OCI registry as an

Reviewer: ```suggestion
Matching? Let's take the GitHub OCI registry as an
```

Author: done

example. There are different owners for

> different repository path (the GitHub org/user).

Reviewer: ```suggestion
different repository paths (the GitHub org/user).
```

Author: done

> Therefore, different credentials needs to be provided

Reviewer: ```suggestion
Therefore, different credentials need to be provided
```

Author: done

for different repository paths.

> For example credentials for ghcr.io/acme can be used

Reviewer: ```suggestion
For example, credentials for ghcr.io/acme can be used
```

Author: done

for a repository ghcr.io/acme/ocm/myimage.

To start with the credentials context we just
provide an explicit mapping for our use case.

First, we create our credentials object as before.

```go
{{include}{../../03-working-with-credentials/02-basic-credential-management.go}{new credentials}}
```

Then we determine the consumer id for our use case.
The repository implementation provides a function
for this task. It provides the most common property
set (no repository path) for an OCI based OCM repository.

```go
{{include}{../../03-working-with-credentials/02-basic-credential-management.go}{consumer id}}
```

The used functions above are just convenience wrappers
around the core type ConsumerId, which might be provided
for dedicated repository/consumer technologies.
Everything can be done directly with the core interface
and property name constants provided by the dedicted technologies. 

> Once we have the id we can finnaly set the credentials for this

Reviewer: ```suggestion
Once we have the id we can finaly set the credentials for this
```

Author: done

id.

```go
{{include}{../../03-working-with-credentials/02-basic-credential-management.go}{set credentials}}
```

Now, the context is prepared to provide credentials 
for any usage of our OCI registry

> Lets test, whether it could provide credentials 

Reviewer: ```suggestion
Let's test whether it could provide credentials 
```

Author: done

for storing our component version.

First, we get the repository object for our OCM repository.

```go
{{include}{../../03-working-with-credentials/02-basic-credential-management.go}{get repository}}
```

Second, we determine the consumer id for our intended repository acccess.

> A credential consumer may provide might provide consumer id information

Reviewer: ```suggestion
A credential consumer may provide consumer id information
```

Author: done

for a dedicated sub user context.
This is supported by the OCM repo implementation for OCI registries.
The usage context is here the component name.

```go
{{include}{../../03-working-with-credentials/02-basic-credential-management.go}{get access id}}
```

Third, we ask the credential context for appropriate credentials.
The basic context method `credctx.GetCredentialsForConsumer` returns
a credentials source interface able to provide credentials
for a changing credentials source. Here, we use a convenience
function directly providing a credentials interface for the
actually valid credentials.
An error is only provided if something went wrong while determining
the credentials. Delivering NO credentials is a valid result.

> the returned interface then offers access to the credential properties.

Reviewer: ```suggestion
The returned interface then offers access to the credential properties.
```

Author: done

via various methods.

```go
{{include}{../../03-working-with-credentials/02-basic-credential-management.go}{get credentials}}
```

Now, we can continue with our basic component version composition
from the last example, or we just display the content.

> The following code snipped show the code for the `context` variant

Reviewer: ```suggestion
The following code snipped shows the code for the `context` variant
```

Author: done

creating a new version, the `read` variant just omits the version creation.

> The rest of the example is then the same.

Reviewer: ```suggestion
The rest of the example is the same.
```

Author: done

```go
{{include}{../../03-working-with-credentials/02-basic-credential-management.go}{add version}}
```

Let's verify the created content and list the versions as known from tour 1.
OCI registries do not support component listers, therefore we
just get and describe the actually added version.

```go
{{include}{../../03-working-with-credentials/02-basic-credential-management.go}{show version}}
```

As we can see in the resource list, our image artifact has been
uploaded to the OCI registry as OCI artifact and the access method has be changed
to `ociArtifact`. It is not longer a local blob.

```go
{{include}{../../03-working-with-credentials/02-basic-credential-management.go}{examine cli}}
```

> This resource access points effectively to the same OCI registry,

Reviewer: ```suggestion
This resource access effectively points to the same OCI registry,
```

Author: done

therefore, we need credentials for accessing the content.
An access method also acts as credential consumer, which
tries to get required credentials from the credential context.
Optionally, an access method can act as provider for a consumer id, so that
it is possible to query the used consumer id from the method object.

```go
{{include}{../../03-working-with-credentials/02-basic-credential-management.go}{image credentials}}
```

Because the credentials context now knows the required credentials,
the access method as credential consumer can access the blob.

```go
{{include}{../../03-working-with-credentials/02-basic-credential-management.go}{image access}}
```

### Providing credentials via credential repositories

The OCM toolset embraces multiple storage
backend technologies, for OCM metadata as well
as for artifacts described by a component version. 
All those technologies typically have their own
way to configure credentials for command line
tools or servers.

The credential management provides so-called
credential repositories. Such a repository
is able to provide any number of named
credential sets. This way any special
credential store can be connected to the

> OCM credential management judt by providing

Reviewer: ```suggestion
OCM credential management just by providing
```

Author: done

an own implementation for the repository interface.

> One such case is the docker config json, a config

Reviewer: ```suggestion
One such case is the Docker config json, a config
```

Author: done

file used by <code>docker login</code> to store
credentials for dedicated OCI registries.

We start again by providing access to the
OCM context and the connected credential context.

```go
{{include}{../../03-working-with-credentials/03-credential-repositories.go}{context}}
```

In package `.../contexts/credentials/repositories` you can find
packages for predefined implementations for some standard credential repositories,
for example `dockerconfig`.

```go
{{include}{../../03-working-with-credentials/03-credential-repositories.go}{docker config}}
```

There are general credential stores, like a HashiCorp Vault

> or type-specific ones, like the docker config json

Reviewer: ```suggestion
or type-specific ones, like the Docker config json
```

Author: done

> used to configure credentials for the docker client.

Reviewer: ```suggestion
used to configure credentials for the Docker client.
```

Author: done

(working with OCI registries).
Those specialized repository implementations are not only able to
provide credential sets, they also know about the usage context
of the provided credentials.

> Therefore, such repository implementations are able to provide credential

Reviewer: ```suggestion
Therefore, such repository implementations are also able to provide credential
```

Author: done

> mappings for consumer ids, also. This is supported by the credential repository

Reviewer: ```suggestion
mappings for consumer ids. This is supported by the credential repository
```

Author: done

API provided by this library.

The docker config is such a case, so we can instruct the
repository to automatically propagate appropriate the consumer id
mappings. This feature is typically enabled by a dedicated specfication
option.

```go
{{include}{../../03-working-with-credentials/03-credential-repositories.go}{propagation}}
```

Implementations for more generic credential repositories can also use this
feature, if the repository allows adding arbitrary metadata. This is for
example used by the `vault` implementation. It uses dedicated attributes
to allow the user to configure intended consumer id properties.

> now we can just add the repository for this specification to

Reviewer: ```suggestion
Now we can just add the repository for this specification to
```

Author: done
